pkg/skaffold/docker/logger/log.go

```diff
@@ -0,0 +1,138 @@
+/*
+Copyright 2021 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logger
+
+import (
+	"context"
+	"io"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker/tracker"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/graph"
+	logstream "github.com/GoogleContainerTools/skaffold/pkg/skaffold/log/stream"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/output"
+)
+
+type Logger struct {
+	out         io.Writer
+	tracker     *tracker.ContainerTracker
+	colorPicker output.ColorPicker
+	client      docker.LocalDaemon
+	muted       int32
+}
+
+func NewLogger(tracker *tracker.ContainerTracker, cfg docker.Config) (*Logger, error) {
+	cli, err := docker.NewAPIClient(cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &Logger{
+		tracker:     tracker,
+		client:      cli,
+		colorPicker: output.NewColorPicker(),
+	}, nil
+}
+
+func (l *Logger) RegisterArtifacts(artifacts []graph.Artifact) {
+	for _, artifact := range artifacts {
+		l.colorPicker.AddImage(artifact.Tag)
+	}
+}
+
+func (l *Logger) Start(ctx context.Context, out io.Writer) error {
+	if l == nil {
+		return nil
+	}
+
+	l.out = out
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case id := <-l.tracker.Notifier():
+				go l.streamLogsFromContainer(ctx, id)
+			}
+		}
+	}()
+	return nil
+}
+
+func (l *Logger) streamLogsFromContainer(ctx context.Context, id string) {
+	tr, tw := io.Pipe()
+	go func() {
+		err := l.client.ContainerLogs(ctx, tw, id)
+		if err != nil {
+			// Don't print errors if the user interrupted the logs
+			// or if the logs were interrupted because of a configuration change
+			if ctx.Err() != context.Canceled {
+				logrus.Warn(err)
+			}
+		}
+		_ = tw.Close()
+	}()
+
+	formatter := NewDockerLogFormatter(l.colorPicker, l.tracker, l.IsMuted, id)
+	if err := logstream.StreamRequest(ctx, l.out, formatter, tr); err != nil {
+		logrus.Errorf("streaming request: %s", err)
```

Comment:
For log errors in general, is there any requirements for sending any information out via the API Server or using the error API?  I think likely this is fine but wasn't sure what the standards were or expectations of skaffold server users.

Reply:
we don't have any well-defined guidelines around this. in theory it would be nice to have all errors collected in a stream surfaced through our API, but at the moment I believe clients (e.g. the IDEs) just attach to `stderr` and read from that. I added a tracking issue for your other comment to emit an event when we lose logs though, so I think that should be enough for anyone interacting with the API.